tfkbnufft/kbnufft.py

```diff
@@ -200,25 +208,30 @@ def kbnufft_adjoint_for_interpob(y, om):
         Returns:
             tensor: The image after adjoint NUFFT.
         """
-        x = adjkbinterp(y, om, interpob)
-
+        grid_y = adjkbinterp(y, om, interpob)
         scaling_coef = interpob['scaling_coef']
         grid_size = interpob['grid_size']
         im_size = interpob['im_size']
         norm = interpob['norm']
         im_rank = interpob.get('im_rank', 2)
-
-        x = ifft_and_scale_on_gridded_data(
-            x, scaling_coef, grid_size, im_size, norm, im_rank=im_rank, multiprocessing=multiprocessing)
+        ifft_y = ifft_and_scale_on_gridded_data(
+            grid_y, scaling_coef, grid_size, im_size, norm, im_rank=im_rank, multiprocessing=multiprocessing)
 
         def grad(dx):
-            x = scale_and_fft_on_image_volume(
+            # Gradients with respect to off grid signal
+            fft_dx = scale_and_fft_on_image_volume(
                 dx, scaling_coef, grid_size, im_size, norm, im_rank=im_rank)
+            dx_dy = kbinterp(fft_dx, om, interpob)
 
-            y = kbinterp(x, om, interpob)
+            # Gradients with respect to trajectory locations
+            r = [tf.linspace(-im_size[i]/2, im_size[i]/2-1, im_size[i]) for i in range(im_rank)]
+            grid_r = tf.cast(tf.meshgrid(*r, indexing='ij'), y.dtype)[None, ...]
```

Comment:
```suggestion
            grid_r = tf.cast(tf.meshgrid(*r, indexing='ij'), dx.dtype)[None, ...]
```

This is a nitpick, but since you multiply with `dx` eventually, we might just cast to this dtype right away